c++/src/kj/common-test.c++

```diff
@@ -793,31 +793,140 @@ TEST(Common, ArrayAsBytes) {
-  KJ_EXPECT(!(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123, 456})));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123, 321}));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const int>({123}));
-
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) == ArrayPtr<const short>({123, 456}));
-  KJ_EXPECT(!(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123, 456})));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123, 321}));
-  KJ_EXPECT(ArrayPtr<const int>({123, 456}) != ArrayPtr<const short>({123}));
-
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) ==
-             ArrayPtr<const char* const>({"foo", "bar"})));
-  KJ_EXPECT(!(ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo", "bar"})));
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo", "baz"})));
-  KJ_EXPECT((ArrayPtr<const StringPtr>({"foo", "bar"}) !=
-             ArrayPtr<const char* const>({"foo"})));
-
-  // operator== should not use memcmp for double elements.
-  double d[1] = { nan() };
-  KJ_EXPECT(ArrayPtr<double>(d, 1) != ArrayPtr<double>(d, 1));
+enum TestOrdering {
+  UNORDERED,
+  EQUAL,
+  LESS,
+  GREATER,
+  NOTEQUAL,
+};
+
+template<typename A, typename B>
+void verifyEqualityComparisons(A a, B b, TestOrdering ord) {
+  const bool expectedEq = ord == EQUAL;
+  KJ_EXPECT((a ==  b) == expectedEq);
+  KJ_EXPECT((b ==  a) == expectedEq);
+  KJ_EXPECT((a !=  b) == !expectedEq);
+  KJ_EXPECT((b !=  a) == !expectedEq);
+}
+
+template<typename T>
+void strongComparisonsTests(T a, T b, TestOrdering ans3way) {
+  const bool expectedEq = ans3way == EQUAL;
+  const bool expectedLT = ans3way == LESS;
+  verifyEqualityComparisons(a, b, ans3way);
+  KJ_EXPECT((a <=  b) == (expectedEq || expectedLT));
+  KJ_EXPECT((b <=  a) == !expectedLT);
+  KJ_EXPECT((a >=  b) == !expectedLT);
+  KJ_EXPECT((b >=  a) == (expectedEq || expectedLT));
+  KJ_EXPECT((a <   b) == expectedLT);
+  KJ_EXPECT((b <   a) == !(expectedEq || expectedLT));
+  KJ_EXPECT((a >   b) == !(expectedEq || expectedLT));
+  KJ_EXPECT((b >   a) == expectedLT);
+}
+
+template<typename A, typename B>
+struct ArrayComparisonTest {
+  Array<A> left;
+  Array<B> right;
+  TestOrdering expectedResult;
+  ArrayComparisonTest(std::initializer_list<A> left, std::initializer_list<B> right, TestOrdering expectedResult) :
+    left(heapArray(left)), right(heapArray(right)), expectedResult(expectedResult) {}
+
+};
+
+KJ_TEST("ArrayPtr comparators for nullptr type") {
+  struct Test {
+    Array<const int> left;
+    TestOrdering expectedResult;
+    Test(std::initializer_list<const int> left, TestOrdering expectedResult) :
+      left(heapArray(left)), expectedResult(expectedResult) {}
+  };
+  Test testCases[] = {
+    {{}, EQUAL},
+    {{123}, GREATER},
+  };
+
+  for(auto const& testCase : testCases) {
```

Comment:
well, you have only tests here, so I wonder if you just eliminate test struct here and write couple verifies manually.